Add extensibility to PreviewOptions

simison opened the pull request:
Resolves https://github.com/WordPress/gutenberg/issues/25309

Builds from https://github.com/WordPress/gutenberg/pull/36119, which includes work from https://github.com/WordPress/gutenberg/pull/31984.

## What?

Most basic version of extending plugin preview menu; and only that. 

Does not allow replacing editor canvas with custom previews yet.

The extension point also does not allow just any custom HTML within the dropdown. Instead the extension is declarative by nature, restricting to specific props only ([background](https://github.com/WordPress/gutenberg/pull/31984#issuecomment-845712238)). This gives us freedom to change the UI later, and plugin developer wouldn't need to care about menu's internal structure.

#### Example usage:

```jsx
<>
	<PluginPreviewMenu
		name="MyPreview1"
		onClick={ () => {} }
		title="My preview 1"
	/>
	<PluginPreviewMenu
		name="MyPreview2"
		onClick={ () => {} }
		title="My preview 2"
	/>
</>
```

#### Example result:

<img width="264" alt="Screenshot 2023-07-26 at 11 29 52" src="https://github.com/WordPress/gutenberg/assets/87168/960ece98-51bb-4312-b166-27cee63646b7">


#### Single item example

<img width="262" alt="Screenshot 2023-07-26 at 11 12 53" src="https://github.com/WordPress/gutenberg/assets/87168/a37bd8e6-43c2-4f5e-a587-2f5dda9e7faf">


### API allows:

- Adding new menu items in the dropdown of post/page editor, within a new menu group, with custom title.
- Using `onClick` handler to open a modal for example.

### API does not allow:

- Adding content to other editors, such as site-editor or widgets.
- Adding custom components/HTML within the dropdown
- Adding multiple menu groups.
- Setting the new item as "selected"; this will make sense once we allow either:
   - replacing the canvas with entirely custom preview,
   - custom canvas preview sizes, or
   - groupings independent from Desktop/Tablet/Mobile selections that can be selected within that group (screenshot below)
- Attaching icon to the menu item; since items cannot be selected, no point providing icon just yet.
- "New tab" functionality which handles saving the draft automatically.
- Plugin branding. I considered always showing icon when defined, but this could cause trouble with showing the selection icon later on.

## Future iterations

### Allow new tabs

Example API could just require adding `href="#"` and we would handle it as new tab automatically:

```jsx
<PluginPreviewMenu
	name="MyPluginPreview"
	href="/my-plugin-preview"
	title="My plugin preview in new tab"
/>
```

Resulting:

<img width="271" alt="Screenshot 2023-07-26 at 11 05 22" src="https://github.com/WordPress/gutenberg/assets/87168/2ce7b80d-9734-4064-9f68-c4250949fa9f">

### Allow selecting items

Allowing selecting (or "setting device") would work well with conjuction in allowing custom canvas preview, or custom canvas preview sizes. https://github.com/WordPress/gutenberg/pull/36119 has one technical implementation idea.

Example API could look like:

```jsx
<PluginPreviewMenu
	icon={ globe }
	name="MyPluginDevice"
	title="My plugin preview"
	setDeviceOnClick
/>
```

Resulting:

<img width="262" alt="Screenshot 2023-07-26 at 11 04 44" src="https://github.com/WordPress/gutenberg/assets/87168/a3a906c7-cd14-49a2-af91-897b924565a9">

Meanwhile current API format would continue supporting e.g. popups:

```jsx
<PluginPreviewMenu
	name="MyPluginPreviewPopup"
	onClick={ () => {} }
	title="My preview popup"
/>
```

### Allow new groupings

Example API could look like:

```jsx
<>
	<PluginPreviewMenu
		icon={ globe }
		group="MyPluginMembersPreview"
		name="MyPluginPaidMembers"
		onClick={ () => {} }
		title="Paid members"
	/>
	<PluginPreviewMenu
		icon={ globe }
		group="MyPluginMembersPreview"
		name="MyPluginFreeMembers"
		onClick={ () => {} }
		title="Free members"
	/>
</>
```

Resulting:

<img width="267" alt="Screenshot 2023-07-26 at 11 07 49" src="https://github.com/WordPress/gutenberg/assets/87168/a68d4119-d171-4bb3-8894-70fdeb673559">

Multiple groupings combined with selection could allow something more sophisticated such as:

<img width="256" alt="Screenshot 2023-07-26 at 11 08 25" src="https://github.com/WordPress/gutenberg/assets/87168/57da36ab-9fc7-41a5-9f0f-f5b2f079752d">



<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

<!-- In a few words, what is the PR actually doing? -->

## Why?
See issue https://github.com/WordPress/gutenberg/issues/25309

Extending the preview menu allows variety of publishing flows and tools, some examples:

- AMP format previews
- Social media share previews
- Previewing same content in different formats such as website, email, RSS, printed
- Displaying with different membership or paid access restrictions/without
- Dark mode for websites that implement it (while editor stays light mode by default)
- Additional display sizes for more exotic targets such as televisions, watches

An example of more complex preview capabilities from Substack:

https://github.com/WordPress/gutenberg/assets/87168/f5004855-fb42-4fe9-8fd0-e143a07e0191

## How?
Extension point utilizing slot/fill.

I didn't add tests yet; I will add those once we are certain this API is acceptable first step.

## Testing Instructions
Test with:

```jsx
 import { PluginPreviewMenu } from '@wordpress/interface';
 import { registerPlugin } from '@wordpress/plugins';

 const MyPreviewMenu = () => (
	<PluginPreviewMenu
		name="MyPreview1"
		onClick={ () => {} }
		title="My preview 1"
	/>
);

registerPlugin( 'my-preview-menu', { render: MyPreviewMenu } );
```

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

mtias commented on packages/block-editor/src/components/preview-options/index.js:
```diff
@@ -56,28 +63,24 @@ export default function PreviewOptions( {
+							</MenuItem>
+						) ) }
 					</MenuGroup>
+					<PluginPreview.Slot
```
We should probably call this `PluginPreviewMenu` as the canvas also needs extensibility points and it'd be confusing?

simison replied:
Agreed! The older PR also [had feedback](https://github.com/WordPress/gutenberg/pull/31984#issuecomment-845721683) that this name was too generic.

mtias commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -366,72 +389,74 @@ export default function VisualEditor( { styles } ) {
-										<LayoutStyle css={ alignCSS } />
-									) }
-									{ postContentLayoutStyles && (
+					<PluginPreviewContent>
```
I'm not sure about this as a wrapper, it seems a bit overbearing. cc @youknowriad

simison replied:
I've left the content wrapper for another iteration.

mtias commented:
Can we split this in two? I think part of the reason it has stalled is that extending the menu is pretty simple, but extending the canvas needs more thought. It'd be good to introduce the first separately.

simison commented on packages/block-editor/src/components/preview-options/index.js:
```diff
@@ -45,6 +46,12 @@ export default function PreviewOptions( {
 		desktop,
 	};
 
+	const devices = [
```
See what term fits best here; there's `deviceType` in many places, and here I shortened it to `device`.

simison committed:
Allow extending preview menu

simison commented:
> Can we split this in two? I think part of the reason it has stalled is that extending the menu is pretty simple, but extending the canvas needs more thought. It'd be good to introduce the first separately.

@mtias agreed, and I think since previews could be done in various ways, we should also approach expanding the API step-by-step. I've updated the code to be very minimal and restricted, yet still quite powerful.

I also elaborated on future ways to expand the API with screenshots.

Ready for review!

youknowriad commented on packages/block-editor/package.json:
```diff
@@ -51,6 +51,7 @@
 		"@wordpress/html-entities": "file:../html-entities",
 		"@wordpress/i18n": "file:../i18n",
 		"@wordpress/icons": "file:../icons",
+		"@wordpress/interface": "file:../interface",
```
I don't think the block editor should depend on "interface", interface is a package to build wp-admin pages (post editor, site editor...) and block editor is a reusable package that can be seen as "TinyMCE" for block editors.

youknowriad commented on packages/block-editor/src/components/preview-options/index.js:
```diff
@@ -79,6 +80,9 @@ export default function PreviewOptions( {
 					</MenuGroup>
+					<PluginPreviewMenu.Slot>
+						{ ( fills ) => <MenuGroup>{ fills }</MenuGroup> }
+					</PluginPreviewMenu.Slot>
```
I'm guessing the reason you depend on interface is because the "PreviewOptions" is placed within the block editor. That was a mistake IMO, I've tried to fix it a long time ago (sometimes duplicating code is better than sharing code) https://github.com/WordPress/gutenberg/pull/30906 but I closed that PR because the added value for that refactoring was a bit small. Potentially today, we could revive that PR (aka remove the shared PreviewOptions component and move it to edit-site and edit-post instead)

youknowriad commented on packages/interface/src/components/plugin-preview-menu/index.js:
```diff
@@ -0,0 +1,58 @@
+const PluginPreviewMenu = ( { name, onClick, title, ...props } ) => (
+	<Fill>
+		<MenuItem
+			className="block-editor-post-preview__button-resize"
```
This classname doesn't follow our guidelines, the component is within "interface" (which can be acceptable) but the classname is like a "block editor" one.

youknowriad commented:
I really appreciate the examples and the thoughts in the descriptions, it all sounds reasonable, the main problem now is the dependency between block-editor and interface.

simison commented:
Thanks for the review @youknowriad ! Happy to continue work on this.

Did I understand you right; it would be better if I moved the component to `block-editor` package directly? I'm not super familiar with the packages and their semantics so happy to just follow your advice.

It's been a while so I don't remember my thinking around `interface` package. It might've come from building on top of other folks, and I vaguely remember some convo about location in https://github.com/WordPress/gutenberg/pull/36119 or https://github.com/WordPress/gutenberg/pull/31984. Anyway, let's find the best spot that works _today_. :-)

youknowriad commented:
I guess IMO, the ideal would be to move the "PreviewOptions" to `edit-post` and `edit-site` (like I did in https://github.com/WordPress/gutenberg/pull/30906) which would create some duplication but IMO is fine. and then add the extensibility to the edit-post version. I'm not sure the component as it exists today adds any value.

simison commented:
@youknowriad @mtias Now that the editor wrapper has been refactored in Gutenberg (which Riad [mentioned](https://github.com/WordPress/gutenberg/pull/50605#issuecomment-1676865460) should happen), it's a good time to resume allowing the preview dropdown to be extended.

Let me know if the premise of this PR needs updating or if the timing isn't great from the Gutenberg team's perspective. Otherwise, we'll proceed with @pkuliga to draft a proposal for:

- Allow expanding the dropdown in the post editor in a declarative way. Page & site editor can be done later in a separate PR.
- Extending creates a new section within the dropdown
- Don't allow expanding the canvas at this time
- Don't allow custom components in the menu
- Don't allow expanding or modifying the device sizes/definitions

So basically minimal, less flexible way to get started and see how that'll get adopted first.

I elaborated in the PR description earlier on the do/don'ts of the API and its potential future direction.

youknowriad commented:
Sounds good to me. I see a few developer related APIs planned for WP 6.7, this would be a good addition as well IMO.

mtias commented:
Yes, let's keep this minimal while we learn how it's used.

simison commented:
@youknowriad @mtias we can continue in https://github.com/WordPress/gutenberg/pull/64644

youknowriad commented:
closing in favor of the other PR

gziolo commented:
Implemented with #64644.

gziolo commented:
Although looking at the code, I figured out that this one was more complex as it also tried to integrate more deeply into the editor screen by allowing to replace the visual code editor.

simison commented:
@gziolo there are two main aspects:

- Extending the preview canvas
- Extending the preview dropdown

There's plenty to figure out on both across different block editors, so [it made sense](https://github.com/WordPress/gutenberg/pull/50605#issuecomment-1651300326) to focus on just dropdown, and keep that aspect pretty limited too.

I've elaborated in [PR description](https://github.com/WordPress/gutenberg/pull/50605#issue-1708153528) (see _"Future iterations"_) potential future directions for the API.

gziolo commented:
> I've elaborated in https://github.com/WordPress/gutenberg/pull/50605#issue-1708153528 (see "Future iterations") potential future directions for the API.

Do we have an issue where we track the future iterations? "Extending the preview canvas" sounds like something that would want to discuss next. Although, that was never considered as an extensibility point despite the fact that there is a way for switching from More Menu between the visual and code editor.

simison commented:
Moved follow-ups to an issue:

- https://github.com/WordPress/gutenberg/pull/64644
